pyro/infer/trace_elbo.py

```diff
@@ -9,6 +8,23 @@
 from pyro.infer.elbo import ELBO
 from pyro.poutine.util import prune_subsample_sites
 from pyro.util import check_model_guide_match, check_site_shape, is_nan
+from pyro.infer.util import MultiFrameTensor
+
+
+# TODO cache some of these computations
+def compute_site_log_r(model_trace, guide_trace, target_site):
+    log_r = MultiFrameTensor()
+    stacks = model_trace.graph["iarange_info"]['iarange_stacks']
+    for name, model_site in model_trace.nodes.items():
+        if model_site["type"] == "sample":
+            log_r_term = model_site["batch_log_pdf"]
+            if not model_site["is_observed"]:
+                guide_site = guide_trace.nodes[name]
+                guide_log_pdf, _, _ = guide_site["score_parts"]
+                log_r_term -= guide_log_pdf
```

Comment:
avoid in-place operation; instead
```py
log_r_term = log_r_term - guide_site["batch_log_pdf"]
```
(also it's a tad simpler to use `site["batch_log_pdf"]` than `site["score_parts"][0]`)

Reply:
done